src/Appwrite/Platform/Workers/Messaging.php

```diff
@@ -179,13 +181,29 @@ private function processMessage(Database $dbForProject, Document $message): void
                         $data = match ($provider->getAttribute('type')) {
                             MESSAGE_TYPE_SMS => $this->buildSMSMessage($messageData, $provider),
                             MESSAGE_TYPE_PUSH => $this->buildPushMessage($messageData),
-                            MESSAGE_TYPE_EMAIL => $this->buildEmailMessage($messageData, $provider),
+                            MESSAGE_TYPE_EMAIL => $this->buildEmailMessage($dbForProject, $messageData, $provider),
                             default => throw new Exception(Exception::PROVIDER_INCORRECT_TYPE)
                         };
 
                         try {
-                            $adapter->send($data);
-                            $deliveredTotal += \count($batch);
+                            $response = new Response($provider->getAttribute('type'));
+                            $response->fromArray($adapter->send($data));
+
+                            $deliveredTotal += $response->getDeliveredTo();
+                            $details[] = $response->getDetails();
+                            foreach ($details as $detail) {
+                                if ($detail['status'] === 'failure') {
+                                    $deliveryErrors[] = `Failed sending to target {$detail['recipient']} with error: {$detail['error']}`;
+                                }
+
+                                // Deleting push targets when token has expired.
+                                if ($detail['error'] === 'Expired device token.') {
```

Comment:
Let's not use a magic string here.

Reply:
We're removing this block altogether as part of https://github.com/appwrite/appwrite/pull/7394